File: app/src/main/java/com/example/workoutapplication/ExerciseManagementAdapter.java
package com.example.workoutapplication;

import androidx.recyclerview.widget.RecyclerView;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

public class ExerciseManagementAdapter extends RecyclerView.Adapter<ExerciseManagementAdapter.ExerciseViewHolder> {

    private Exercise[] localDataSet;

    // Providing a reference to type of views being used
    // (custom ViewHolder)
    public static class ExerciseViewHolder extends RecyclerView.ViewHolder {
        private final ImageView exerciseThumbnail;
        private final TextView exerciseName;
        private final TextView exerciseDescription;

        // ViewHolder constructed using a given view (which it will hold)
        public ExerciseViewHolder(View view) {
            super(view);

            // NOTE: could define click listener here

            // in the view, find the element with id textView
            exerciseThumbnail = (ImageView) view.findViewById(R.id.iv_exercise_thumbnail);
            exerciseName = (TextView) view.findViewById(R.id.tv_exercise_name);
            exerciseDescription = (TextView) view.findViewById(R.id.tv_exercise_description);
        }

        public ImageView getExerciseThumbnailView() {
            return exerciseThumbnail;
        }
        public TextView getExerciseNameView() {return exerciseName;}
        public TextView getExerciseDescription() {return exerciseDescription;}
    }

    /**
     * Initialize the dataset of the adapter
     *
     * @param dataSet String[] containing the data to populate views
     *                to be used by RecyclerView
     */
    public ExerciseManagementAdapter(Exercise[] dataSet) {
        localDataSet = dataSet;
    }

    // Create new views (will be invoked by the layout manager
    @Override
    public ExerciseViewHolder onCreateViewHolder(ViewGroup viewGroup, int viewType) {
        // Create a new view, which defines the UI of the list item
        View view = LayoutInflater.from(viewGroup.getContext())
                .inflate(R.layout.exercise_row_item, viewGroup, false);

        return new ExerciseViewHolder(view);
    }

    // Replace the contents of a view (will be invoked by layout manager)
    // Alternatively: Adds/changes data in the ViewHolder (different views!)
    @Override
    public void onBindViewHolder(ExerciseViewHolder viewHolder, final int position) {

        // Get element from your dataset at this position and
        // replace the contents of the view with that element
        viewHolder.getExerciseNameView().setText(localDataSet[position].getName());
        viewHolder.getExerciseDescription().setText(localDataSet[position].getDescription());
    }

    @Override
    public int getItemCount() {
        return localDataSet.length;
    }

}

File: app/src/main/java/com/example/workoutapplication/ExerciseManagementAdapter.kt
package com.example.workoutapplication

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.example.workoutapplication.ExerciseManagementAdapter.ExerciseViewHolder
import com.example.workoutapplication.dataClasses.Exercise

// TODO force any activity using this adapter to ensure it implements interface ExerciseRecyclerViewListener
class ExerciseManagementAdapter (private val localDataSet: ArrayList<Exercise>,
                                 private val listener: ExerciseRecyclerViewListener)
    : RecyclerView.Adapter<ExerciseViewHolder>() {

    interface ExerciseRecyclerViewListener {
        fun onListItemClick(position: Int)
    }

    // Providing a reference to type of views being used
    // (custom ViewHolder)
    inner class ExerciseViewHolder(view: View) : RecyclerView.ViewHolder(view),
        View.OnClickListener {
            val exerciseThumbnailView: ImageView
            val exerciseNameView: TextView
            val exerciseDescription: TextView

            // ViewHolder constructed using a given view (which it will hold)
            init {
                // When this view (list item) is clicked, this class will be returned
                // as an interface View.OnClickListener, which can then run onClick()
                view.setOnClickListener(this)

                // in the view, find the element with id textView
                exerciseThumbnailView = view.findViewById(R.id.iv_exercise_thumbnail)
                exerciseNameView = view.findViewById(R.id.tv_exercise_name)
                exerciseDescription = view.findViewById(R.id.tv_exercise_description)
            }

            override fun onClick(v: View?) {
                val position = adapterPosition
                // NO_POSITION is returned if the user clicks a
                // removed item while the animation is running
                if (position != RecyclerView.NO_POSITION)
                    listener.onListItemClick(position)
            }
    }

    // Create new views (will be invoked by the layout manager
    override fun onCreateViewHolder(viewGroup: ViewGroup, viewType: Int): ExerciseViewHolder {
        // Create a new view, which defines the UI of the list item
        val view = LayoutInflater.from(viewGroup.context)
            .inflate(R.layout.exercise_row_item, viewGroup, false)
        return ExerciseViewHolder(view)
    }

    // Replace the contents of a view (will be invoked by layout manager)
    // Alternatively: Adds/changes data in the ViewHolder (different views!)
    override fun onBindViewHolder(viewHolder: ExerciseViewHolder, position: Int) {

        // Get element from your dataset at this position and
        // replace the contents of the view with that element
        viewHolder.exerciseNameView.text = localDataSet[position].name
        viewHolder.exerciseDescription.text = localDataSet[position].description
    }

    override fun getItemCount(): Int {
        return localDataSet.size
    }
}